Adding Spring Boot enhancements, Sring Data Repository, Testcontainers

artur-ciocanu opened the pull request:
# Description

This PR adds support for:
- Spring Boot Starters
- Spring Data Repository
- Spring Boot Testcontainers

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1088 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

salaboy commented:
This depends on this: https://github.com/dapr/java-sdk/pull/1085 so please help us to merge that one first

salaboy commented on sdk-springboot/pom.xml:
```diff
@@ -10,15 +10,27 @@
+  <groupId>io.dapr.spring</groupId>
+  <artifactId>dapr-sdk-spring-parent</artifactId>
+  <packaging>pom</packaging>
   <version>1.12.0-SNAPSHOT</version>
```
@artur-ciocanu we need to change this to 0.12.0-SNAPSHOT (we use the 0 to show that this artifact is not stable yet.

artur-ciocanu replied:
Done

salaboy commented on sdk-springboot/pom.xml:
```diff
@@ -10,15 +10,27 @@
-    <maven.deploy.skip>false</maven.deploy.skip>
     <springboot.version>3.2.6</springboot.version>
+    <testcontainers-dapr.version>0.10.14</testcontainers-dapr.version>
+    <testcontainers-test.version>1.19.8</testcontainers-test.version>
```
@artur-ciocanu this was now moved to the parent pom, so it should be ok to remove it from here when the testcontainers PR gets merged.

artur-ciocanu replied:
Right, once Testcontainers support is in I will adjust

salaboy commented on sdk-springboot/pom.xml:
```diff
@@ -10,15 +10,27 @@
   <properties>
-    <maven.deploy.skip>false</maven.deploy.skip>
     <springboot.version>3.2.6</springboot.version>
+    <testcontainers-dapr.version>0.10.14</testcontainers-dapr.version>
```
@artur-ciocanu We will not need this one when the PR is merged.

artur-ciocanu replied:
True, I will remove once Testcontainers PR is in.

salaboy commented on sdk-springboot/pom.xml:
```diff
@@ -10,15 +10,27 @@
     <springboot.version>3.2.6</springboot.version>
+    <testcontainers-dapr.version>0.10.14</testcontainers-dapr.version>
+    <testcontainers-test.version>1.19.8</testcontainers-test.version>
+    <maven.compiler.source>17</maven.compiler.source>
```
@artur-ciocanu I think we can remove this one too.. as this is all coming from the parent pom?

artur-ciocanu replied:
Parent POM enforces Java 8 and compiles using Java 11. We need this explicitly.

salaboy commented on sdk-springboot/sdk-spring-core/pom.xml:
```diff
@@ -0,0 +1,55 @@
+    <!-- https://mvnrepository.com/artifact/com.mysql/mysql-connector-j -->
+    <dependency>
+      <groupId>com.mysql</groupId>
+      <artifactId>mysql-connector-j</artifactId>
```
@artur-ciocanu why do we need this? We are not supposed to be accessing MySQL directly, even for tests.

artur-ciocanu replied:
Great question! So I had a chat with Testcontainers folks on their Slack and I have found the following:
- PostgreSQL and MySQL both derive from `JdbcDatabaseContainer`, so potentially to use any of these containers you need a JDBC driver.
- A container can provide its own custom so called `WaitStrategy` and MySQL uses the default one based on a SQL query, while PostgreSQL uses one based on a simple log message.

Hence for MySQL we need a JDBC driver, if it is critical, we can implement our own MySQL container and provide a custom wait strategy similar to the one used by PostgreSQL.

salaboy replied:
@artur-ciocanu I think all we need is a comment in the XML why the dependency is there.. so we know it is on purpose

salaboy commented on sdk-springboot/sdk-spring-core/src/main/java/io/dapr/spring/data/AbstractDaprKeyValueAdapter.java:
```diff
@@ -0,0 +1,139 @@
+
+  @Override
+  public void clear() {
+    throw new UnsupportedOperationException("'clear' method is not supported");
```
@artur-ciocanu shoudn't this do nothing instead of throwing an exception? I would rather have a `do nothing` comment here. Imagine that you have an application that is using the Spring Data Redis Keyvalue implementation and they are already calling clear, when they change to the Dapr implementation, they will get an exception and they will need to change their code.

artur-ciocanu replied:
Fixed.

salaboy commented on sdk-springboot/sdk-spring-core/src/main/java/io/dapr/spring/data/AbstractDaprKeyValueAdapter.java:
```diff
@@ -0,0 +1,139 @@
+
+  @Override
+  public CloseableIterator<Map.Entry<Object, Object>> entries(String keyspace) {
+    throw new UnsupportedOperationException("'entries' method is not supported");
```
@artur-ciocanu we might need to implement this one.. I am not even sure why this is needed.. but it might be worth trying to implement this one

artur-ciocanu replied:
This is harder to implement and to be honest it is very KB specific.

salaboy commented on sdk-springboot/sdk-spring-core/src/main/java/io/dapr/spring/data/repository/query/DaprPredicateQueryCreator.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.function.Predicate;
+
+/**
+ * This class is copied from https://github.com/spring-projects/spring-data-keyvalue/blob/ff441439124585042dd0cbff952f977a343444d2/src/main/java/org/springframework/data/keyvalue/repository/query/PredicateQueryCreator.java#L46
```
@artur-ciocanu can we remove this?

salaboy commented on sdk-springboot/sdk-spring-core/src/test/java/io/dapr/spring/data/AbstractBaseIT.java:
```diff
@@ -0,0 +1,25 @@
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+@Testcontainers
+public abstract class AbstractBaseIT {
```
@artur-ciocanu we will need to move this to `sdk-tests` there is an IT package. Look at how I moved the test in the testcontainers-dapr module -> https://github.com/dapr/java-sdk/pull/1085/files#diff-3567b77e1a2a23e9200900907bd789529d99ccf8968b54fd0f8fc8eac3c63d24

artur-ciocanu replied:
Moved

salaboy commented on sdk-springboot/sdk-spring-core/src/test/java/io/dapr/spring/data/AbstractPostgreSQLBaseIT.java:
```diff
@@ -0,0 +1,72 @@
+limitations under the License.
+*/
+
+package io.dapr.spring.data;
```
@artur-ciocanu same here.. we need to move this to `sdk-tests`

artur-ciocanu replied:
Moved

salaboy commented on sdk-springboot/sdk-spring-core/src/test/java/io/dapr/spring/data/MySQLDaprKeyValueTemplateIT.java:
```diff
@@ -0,0 +1,229 @@
+limitations under the License.
+*/
+
+package io.dapr.spring.data;
```
@artur-ciocanu same here.. `sdk-tests`

artur-ciocanu replied:
Moved

salaboy commented on sdk-springboot/sdk-spring-core/src/test/java/io/dapr/spring/data/repository/DaprKeyValueRepositoryIT.java:
```diff
@@ -0,0 +1,123 @@
+limitations under the License.
+*/
+
+package io.dapr.spring.data.repository;
```
@artur-ciocanu same.. move to IT tests in `sdk-tests`

salaboy commented on sdk-springboot/sdk-springboot-autoconfigure/src/test/java/io/dapr/spring/boot/autoconfigure/BaseIntegrationTest.java:
```diff
@@ -0,0 +1,45 @@
+limitations under the License.
+*/
+
+package io.dapr.spring.boot.autoconfigure;
```
@artur-ciocanu same.. needs to be moved to `sdk-tests`

artur-ciocanu replied:
Moved

salaboy commented on sdk-springboot/sdk-springboot-autoconfigure/src/test/java/io/dapr/spring/boot/autoconfigure/pubsub/DaprPubSubAutoConfigurationIT.java:
```diff
@@ -0,0 +1,70 @@
+limitations under the License.
+*/
+
+package io.dapr.spring.boot.autoconfigure.pubsub;
```
@artur-ciocanu same `sdk-tests`

artur-ciocanu replied:
Moved

salaboy commented on sdk-springboot/sdk-springboot-autoconfigure/src/test/java/io/dapr/spring/boot/autoconfigure/pubsub/TestRestController.java:
```diff
@@ -0,0 +1,52 @@
+    return "OK";
+  }
+
+  @Topic(name = topicName, pubsubName = pubSubName)
```
@artur-ciocanu this is interesting.. this @Topic annotation needs to be renamed to @Listener or @DaprListener to follow other spring project name conventions. This breaks compatibility so, I will defer to @artursouza recommendation here. But I think this change needs to be included in this PR. Maybe we can have both @DaprListener and @Topic while we mark @Topic as deprecated.

artur-ciocanu replied:
We will handle this later

salaboy commented on sdk-springboot/sdk-springboot-starters/sdk-springboot-starter/pom.xml:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
@artur-ciocanu to your and @ThomasVitale  point .. I think we need to remove `sdk` from all the directories under the sdk-springboot parent one.

salaboy commented on dapr-spring/dapr-spring-boot-autoconfigure/src/main/java/io/dapr/spring/boot/autoconfigure/statestore/DaprStateStoreAutoConfiguration.java:
```diff
@@ -0,0 +1,62 @@
+@AutoConfiguration(after = DaprClientAutoConfiguration.class)
+@ConditionalOnClass({DaprClient.class, KeyValueAdapter.class})
+@EnableConfigurationProperties(DaprStateStoreProperties.class)
+public class DaprStateStoreAutoConfiguration {
```
@ThomasVitale @artur-ciocanu I am starting to think that we don't need this autoconfiguration. Mostly because if you look at Spring Data, you will see that users need to basically provide a @configuration class that creates the instances of the KeyValueAdapterResolver and the KeyValueTempplate -> https://docs.spring.io/spring-data/keyvalue/reference/keyvalue/template.html . So maybe in our attempt to make things easier, we made it more complicated for ourselves. As soon as we have in our Docs that whenever you want a DaprKeyValue template, you just need to provide a @Configuration class to get the instances based on the DaprBuilder and the DaprStateProperties. What do you think folks?

artur-ciocanu replied:
As I mentioned in a previous comment, I have moved the `org.springframework.boot.autoconfigure.AutoConfiguration.imports` to starter.

salaboy replied:
Hmm.. I am not sure what this means.. because I think we don't even need the AutoConfigurations.. Can you share the link of where you moved things? I need more context to understand what you meant by:
```
 have moved the org.springframework.boot.autoconfigure.AutoConfiguration.imports to starter.
```

salaboy commented on sdk-workflows/pom.xml:
```diff
@@ -1,6 +1,6 @@
 <project
-    xmlns="http://maven.apache.org/POM/4.0.0"
```
@artur-ciocanu there is no change in here.. can we make sure we don't change this file?

artur-ciocanu replied:
Reverted

salaboy commented on dapr-spring/dapr-spring-boot-testcontainers/src/test/java/io/dapr/testcontainers/module/DaprModuleTests.java:
```diff
@@ -0,0 +1,87 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+@SpringBootTest(classes = MyTestWithWorkflowsApplication.class, webEnvironment = WebEnvironment.RANDOM_PORT)
```
@artur-ciocanu I have the feeling that we will need to move these tests to sdk-tests.. because they start containers.

artur-ciocanu replied:
Moved

salaboy commented on dapr-spring/dapr-spring-boot-autoconfigure/src/main/java/io/dapr/spring/boot/autoconfigure/pubsub/DaprPubSubAutoConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+@EnableConfigurationProperties(DaprPubSubProperties.class)
+public class DaprPubSubAutoConfiguration {
+
+  @Bean
```
@artur-ciocanu @ThomasVitale the more I look into this.. I believe we don't need it.. this is @Configuration that the user should provide. -> https://docs.spring.io/spring-kafka/reference/kafka/sending-messages.html To quote Spring Kafka docs: 
```
To use the template, you can configure a producer factory and provide it in the template’s constructor. The following example shows how to do so:

@Bean
public ProducerFactory<Integer, String> producerFactory() {
    return new DefaultKafkaProducerFactory<>(producerConfigs());
}

@Bean
public Map<String, Object> producerConfigs() {
    Map<String, Object> props = new HashMap<>();
    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
    // See https://kafka.apache.org/documentation/#producerconfigs for more properties
    return props;
}

@Bean
public KafkaTemplate<Integer, String> kafkaTemplate() {
    return new KafkaTemplate<Integer, String>(producerFactory());
}
```

salaboy replied:
@artur-ciocanu My recommendation will be to remove the PubSub and KeyvalueTemplate autoconfigurations, and have tests that shows that if the user adds the right beans when they want to use the messagingTemplate and the keyvalueTemplate @beans everything works. 
For @EnableDaprRepositories is different because we are explicitly enabling a feature that comes with default keyvaluetemplate instances.

artur-ciocanu replied:
I have moved the `org.springframework.boot.autoconfigure.AutoConfiguration.imports` to starter and this should resolve the issue. The starter is not included in the `sdk-tests`.

salaboy commented on dapr-spring/dapr-spring-boot-starters/dapr-spring-boot-starter/pom.xml:
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
@artur-ciocanu do we need this?

salaboy replied:
@artur-ciocanu  oh.. ok.. so maybe this is where we create the spring boot starters for Data and Messaging which shows how all comes together with the right dependencies. I think that here is also the place to have a combined starter for all Dapr submodules. This would be the place to have tests to demonstrate what the user should provide and what will be automatically wired by the spring boot autoconfiguration for the Dapr client.

artur-ciocanu replied:
Yes exactly, we use starters to bootstrap auto-configurations. We will have one that bundles all the beans, then we can have one for `data`, one for `messaging`, etc.

salaboy commented on dapr-spring/spotbugs-exclude.xml:
```diff
@@ -0,0 +1,11 @@
+<FindBugsFilter>
+    <Match>
+        <Package name="~io\.dapr\.spring.*"/>
```
@artur-ciocanu is this temporary? what is this excluding exactly?

artur-ciocanu replied:
These are removing the complains that we are passing DaprClient and things like that in the `DaprKeyValueTemplate` for example.

artursouza replied:
Can you bring this back, please? We should look into those. Findbugs catches some nice stuff.

artur-ciocanu replied:
I have provided the explanations in another comment.

salaboy commented on dapr-spring/pom.xml:
```diff
@@ -0,0 +1,174 @@
+    <testcontainers-dapr.version>0.12.0-SNAPSHOT</testcontainers-dapr.version>
+    <testcontainers-test.version>1.19.8</testcontainers-test.version>
+    <dapr.sdk.version>1.12.0-SNAPSHOT</dapr.sdk.version>
+    <dapr-workflows.sdk.version>0.12.0-SNAPSHOT</dapr-workflows.sdk.version>
```
@artur-ciocanu I think in master now we have 
`<dapr.sdk.alpha.version>0.12.0-SNAPSHOT</dapr.sdk.alpha.version>` -> https://github.com/dapr/java-sdk/blob/master/pom.xml#L20C5-L20C69

artur-ciocanu replied:
Fixed.

salaboy commented on dapr-spring/dapr-spring-boot-starters/dapr-spring-boot-starter/src/main/resources/META-INF/spring/org.springframework.boot.autoconfigure.AutoConfiguration.imports:
```diff
@@ -0,0 +1,3 @@
+io.dapr.spring.boot.autoconfigure.client.DaprClientAutoConfiguration
+io.dapr.spring.boot.autoconfigure.pubsub.DaprPubSubAutoConfiguration
```
@artur-ciocanu ok.. so you meant this.. but I strongly thing that we don't need `io.dapr.spring.boot.autoconfigure.pubsub.DaprPubSubAutoConfiguration` and `io.dapr.spring.boot.autoconfigure.statestore.DaprStateStoreAutoConfiguration`, as I pointed in the docs for spring-data-redis and spring-kafka they ask the users to provide these @Beans in their applications. I think we should do that instead of adding our own AutoConfigurations.

artur-ciocanu replied:
This has been addressed.

salaboy commented on sdk/pom.xml:
```diff
@@ -134,16 +134,6 @@
       <version>1.7</version>
       <scope>test</scope>
     </dependency>
-    <dependency>
```
@artur-ciocanu are these not used?

artur-ciocanu replied:
These are duplicates that Maven was complaining about.

salaboy commented:
@cicoyle @artursouza this test here: https://github.com/dapr/java-sdk/actions/runs/10578048505/job/29307323653?pr=1089
is failing with this error: 

```
MethodInvokeIT.testInvokeTimeout:95 The message contains DEADLINE_EXCEEDED: DEADLINE_EXCEEDED: context deadline exceeded
```

But the assertion is done on contains "CallOptions deadline exceeded after" , which in my opinion is causing the flakiness as Daprd (the runtime) is not consistently returning the same error. 

@cicoyle maybe you can help me to investigate what is the difference between "context deadline exceeded" and "CallOptions deadline exceeded after" 

https://github.com/dapr/java-sdk/blob/master/sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java#L95

artur-ciocanu commented:
> @cicoyle @artursouza this test here: https://github.com/dapr/java-sdk/actions/runs/10578048505/job/29307323653?pr=1089 is failing with this error:
> 
> ```
> MethodInvokeIT.testInvokeTimeout:95 The message contains DEADLINE_EXCEEDED: DEADLINE_EXCEEDED: context deadline exceeded
> ```
> 
> But the assertion is done on contains "CallOptions deadline exceeded after" , which in my opinion is causing the flakiness as Daprd (the runtime) is not consistently returning the same error.
> 
> @cicoyle maybe you can help me to investigate what is the difference between "context deadline exceeded" and "CallOptions deadline exceeded after"
> 
> https://github.com/dapr/java-sdk/blob/master/sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java#L95
> 
> In my experience so far, this test is returning "CallOptions deadline exceeded after" one every 5 attempts.

@salaboy, unless @artursouza and @cicoyle have a different opinion, I think asserting the `code` of `StattusRuntimeException` being equal to enum value `DEADLINE_EXCEEDED(4)` is more than enough.

I have adjusted the test, please take a look and let me know your thoughts.

artur-ciocanu commented:
@artursouza and @cicoyle the build is 🟢. I will need your help with `License Compliance`, I don't think we use anything "dangerous", but we might need your approval.

Thank you!

artursouza commented on .github/workflows/build.yml:
```diff
@@ -117,9 +117,12 @@ jobs:
       uses: codecov/codecov-action@v4.1.0
     - name: Install jars
       run: ./mvnw install -q -B -DskipTests
+    - name: Integration tests with testcontainers using spring boot version ${{ matrix.spring-boot-version }}
```
Why this needs to be separate?

salaboy replied:
@artursouza good question, this helps to isolate tests running with testcontainers from tests running with the local `daprd` binary. Running both concurrently generates issues that are hard to reproduce locally. I think it is sensitive to separate them until we can align both approaches better.

artur-ciocanu replied:
As @salaboy mentioned, when running everything together we have seen pretty weird behavior where GRPC channel is not properly closed and things like that, which lead to errors that are hard to reproduce in isolation. So the only way to overcome this problem was to have two groups of test:
- Testcontainers based - these are ITs annotated with `@Tag("testcontainers")` and are relying on Testcontainers library to start, stop and cleanup any started containers.
- Other ITs - these are ITs that are using `daprd` process with Docker Compose etc

artursouza replied:
This is a problem. It means the tests are leaving something running behind and influencing with other tests. If we merge this, it will be a technical debt. I don't mind having the Tag for people to run them separately locally, but they should run together. It is also more cognitive burden on contributions to figure out that cannot run the ITs together.

artursouza replied:
If this cannot be solved in this PR, I will still merge but I will create an issue for someone to fix this tech debt.

artursouza commented on dapr-spring/dapr-spring-boot-autoconfigure/src/main/java/io/dapr/spring/boot/autoconfigure/client/DaprClientBuilderConfigurer.java:
```diff
@@ -0,0 +1,52 @@
+  private List<DaprClientCustomizer> customizers;
+
+  void setDaprClientCustomizer(List<DaprClientCustomizer> customizers) {
+    this.customizers = customizers;
```
immutable.

artur-ciocanu replied:
Done.

artursouza commented on dapr-spring/dapr-spring-boot-testcontainers/src/main/java/io/dapr/testcontainers/DaprModule.java:
```diff
@@ -0,0 +1,49 @@
+
+  @Container
+  DaprContainer dapr = new DaprContainer("daprio/daprd:1.13.2")
+      .withAppName("local-dapr-app")
```
All these should be coming as param and not hardcoded.

artursouza commented on dapr-spring/dapr-spring-data/src/main/java/io/dapr/spring/data/DaprKeyValueAdapterResolver.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.Set;
+
+public class DaprKeyValueAdapterResolver implements KeyValueAdapterResolver {
+  private static final List<String> MYSQL_VALUES = Arrays.asList("state.mysql-v1", "bindings.mysql-v1");
```
All these should be immutable.

artur-ciocanu replied:
Done

artursouza commented on dapr-spring/spotbugs-exclude.xml:
```diff
@@ -0,0 +1,11 @@
+
+    <Match>
+        <Package name="~io\.dapr\.spring.*"/>
+        <Bug pattern="EI_EXPOSE_REP2"/>
```
Same.

artursouza commented on sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java:
```diff
@@ -88,11 +88,12 @@ public void testInvokeTimeout() throws Exception {
+            Status.Code code = exception.getStatus().getCode();
+
             assertTrue(delay >= TIMEOUT_MS, "Delay: " + delay + " is not greater than timeout: " + TIMEOUT_MS);
-            assertTrue(message.contains("DEADLINE_EXCEEDED"), "The message contains DEADLINE_EXCEEDED: " + message);
```
This was working before, why removed?

salaboy replied:
This is related to this -> https://github.com/dapr/java-sdk/pull/1089#issuecomment-2312502731 That error message if flaky. I don't think we should remove the assertion, but we need to find why Dapr Runtime is returning two different messages, I am not sure to understand the difference between them.

artur-ciocanu replied:
@artursouza I have seen this IT test failing on and off on multiple PRs, just as @salaboy mentioned. In my opinion, relying on exception message leads to flaky tests. In this particular test we are interested in the fact that a timeout has been reached, hence looking for `Status.Code` equal to `DEADLINE_EXCEEDED` should be enough.

artursouza reviewed (changes requested):
Pretty close. Have a few comments. Thanks a lot for your contribution.

artursouza commented:
> @artursouza and @cicoyle the build is 🟢. I will need your help with `License Compliance`, I don't think we use anything "dangerous", but we might need your approval.
> 
> Thank you!

The violation is coming from MySQL Connector/J 8.2.0 Community for using GPL 2.0.

Is there an alternative we can use?

salaboy commented on dapr-spring/dapr-spring-boot-testcontainers/src/main/java/io/dapr/testcontainers/DaprModule.java:
```diff
@@ -0,0 +1,49 @@
+/*
```
@artur-ciocanu are we using this anywhere downstream? maybe we just don't need this, the entire module is just an example and I can see why people will never use this.

artur-ciocanu replied:
I have moved this to `sdk-tests` to keep it as an example.

salaboy commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprContainerTest.java:
```diff
@@ -145,15 +140,20 @@ public void testPlacement() throws Exception {
+
+  private DaprClientBuilder createDaprClientBuilder() {
+    return  new DaprClientBuilder()
+        .withPropertyOverride(Properties.HTTP_PORT, String.valueOf(DAPR_CONTAINER.getHttpPort()))
```
🔥🔥🔥🔥
I love the way that this worked out 👍

artur-ciocanu replied:
Yeyyy

artur-ciocanu commented:
> > @artursouza and @cicoyle the build is 🟢. I will need your help with `License Compliance`, I don't think we use anything "dangerous", but we might need your approval.
> > Thank you!
> 
> The violation is coming from MySQL Connector/J 8.2.0 Community for using GPL 2.0.
> 
> Is there an alternative we can use?

I have created a `CustomMySQLContainer` implementation with a log based wait strategy and `MySQL Connector J` is  no longer required.

artur-ciocanu commented:
@artursouza and @salaboy you can check the SpotBugs failures here: https://github.com/dapr/java-sdk/actions/runs/10601809365/job/29382433463?pr=1089#step:16:3094. This is after I have removed SpotBugs exclude from `dapr-spring` module.

I don't think having those two exclusion rules for SpotBugs will lead to issues. As I already mentioned most of the classes that we inject via dependency injections are Spring Beans and there is nothing we can do about it.

artursouza commented on sdk-tests/src/test/java/io/dapr/it/spring/data/DaprKeyValueRepositoryIT.java:
```diff
@@ -0,0 +1,142 @@
+      .withDatabaseName("dapr_db_repository")
+      .withUsername("postgres")
+      .withPassword("password")
+      .withExposedPorts(5432)
```
This is why it is having problems. The integration tests must use ports detected at runtime and not a fixed one.

eddumelendez replied:
withExposePort is not needed because PostgreSQLContainer already have it. This is not a fixed port exposed in the host, it declared the port exposed from the container and the host port will be a random one.

eddumelendez commented on sdk-tests/src/test/java/io/dapr/it/spring/data/DaprKeyValueRepositoryIT.java:
```diff
@@ -0,0 +1,142 @@
+
+  @DynamicPropertySource
+  static void daprProperties(DynamicPropertyRegistry registry) {
+    DAPR_CONTAINER.start();
```
this is not needed. Test is using Testcontainers JUnit 5 integration so @Testcontainers will start those annotated with @Container

artur-ciocanu replied:
Removed, thank you.

eddumelendez commented on sdk-tests/src/test/java/io/dapr/it/spring/data/DaprKeyValueRepositoryIT.java:
```diff
@@ -0,0 +1,142 @@
+@ContextConfiguration(classes = TestDaprSpringDataConfiguration.class)
+@Testcontainers
+@Tag("testcontainers")
+public class DaprKeyValueRepositoryIT {
```
```suggestion
class DaprKeyValueRepositoryIT {
```

artur-ciocanu replied:
I have followed the pattern used in other integration tests.

eddumelendez commented on sdk-tests/src/test/java/io/dapr/it/spring/data/DaprKeyValueRepositoryIT.java:
```diff
@@ -0,0 +1,142 @@
+  }
+
+  @Test
+  public void testFindById() {
```
```suggestion
  void testFindById() {
```

artur-ciocanu replied:
I have followed the pattern used in `sdk-tests` modules, so I will leave the modifier as is.

artur-ciocanu commented:
> Pretty close. Have a few comments. Thanks a lot for your contribution.

@artursouza thanks a lot for your review. I have addressed or answered all your comments. I have provided an example regarding SpotBugs failures.

If everything looks good, could you please approve and merge. Thank you.

artur-ciocanu commented:
@artursouza @salaboy @cicoyle the build is 🟢 I have addressed all the comments. Whenever you have a chance could you please take a look.

Thank you 🙇

artur-ciocanu commented:
@artursouza I have addressed  all the review comments. The build is 🟢. I am not sure what's wrong with `FOSSA`. Could you please take another look and approve it.

Thank you 🙇 !
